mercury/src/internal/pack/cache_object.rs

```diff
@@ -98,8 +107,11 @@ impl Default for CacheObject {
 // ! the implementation of HeapSize is not accurate, only calculate the size of the data_decompress
 // Note that: mem_size == value_size + heap_size, and we only need to impl HeapSize because value_size is known
 impl HeapSize for CacheObject {
+    /// For [`ObjectType::OffsetDelta`] and [`ObjectType::HashDelta`], 
+    /// `final_size` is the size of the expanded object;
+    /// for other types, `final_size` is 0.
     fn heap_size(&self) -> usize {
-        self.data_decompress.heap_size()
+        self.data_decompress.heap_size() + self.final_size
```

Comment:
`final_size` is the size of final `new` object after `rebuild_delta`, right?

Here, `data_decompress.heap_size()` + `final_size` == the memory usage when both the `delta` object and the `new` object exist (This situation will only occur during the **rebuild** process.). 
Is this intentional? Is it aimed at achieving better memory control?

According to the previous control strategy, only `final_size` or `data_decompress.heap_size()` should be needed.

Reply:
> final_size is the size of final new object after rebuild_delta, right?

It's actaully the size of thr `data_decompress` field of the new object, *i.e.* `heap_size`, but there's no much difference

> Is this intentional? Is it aimed at achieving better memory control?

Yes, for a period of time, in the `rebuild_delta` function, the space allocated for both the old object and the new object exists simultaneously.
``` Rust
    pub fn rebuild_delta(delta_obj: CacheObject, base_obj: Arc<CacheObject>) -> CacheObject {
        // ...
        let mut result = Vec::with_capacity(result_size);  // start from here
        loop {
			// ...
        }
        CacheObject { ... }  // new object returns
    }    // delta object is dropped

```

Reply:
> Yes, for a period of time, in the rebuild_delta function, the space allocated for both the old object and the new object exists simultaneously.

If you did this intentionally, please add a **comment** to explain your reasoning. We must consider maintainability for future developers.

Semantically, having `final_size` = 0 for **undeltified** objects is quite **confusing**. Consider using a different name, such as `delta_final_size`.

Reply:
> Yes, for a period of time, in the rebuild_delta function, the space allocated for both the old object and the new object exists simultaneously.

Additionally, we need to evaluate the impact of this memory control strategy on decode speed.

Reply:
Under the 4GiB memory limit, the peak memory usage of the new implementation 1fb33897a67d90efbd89844947734578992b9b2a is around 4.09GiB, while the old one 414ef7b6873a61c62380da6acbbb88a1b10915c1 was 4.28GiB. However, it turns out that there is little impact on running speed.

Reply:
If you notice that each `CacheObject` in `to_entry` is copied once, and all our current tests simply discard the `Entry`, you can realize that the current memory control will easily fail...

Reply:
Of course, we have been aware of this issue since we introduced the callback function to send objects externally. 

Objects sent externally are **not included** in memory control; we only focus on the memory used by the `decode` function **itself**, not the entire process.

Reply:
I believe more things could be done here,not sure though...

--

Could this conversation be marked solved?

Reply:
The external caller requires the `Entry` type, so we cannot use `Arc<CacheObject>`. 
At the same time, we want to limit the tracking of `CacheObject` to within the decode function. 
Ideally, we could achieve sharing of the underlying `data_decompress(Vec<u8>)` (which does not change for new objects), but this might **complicate** the system.

At current stage, both speed and memory control are **acceptable**. The biggest issue is the unexplained memory not being released (after the `decode` function ends, you can observe using `sleep()` that some memory is not returned to the operating system, which might be related to the memory allocator).

Reply:
> Could this conversation be marked solved?

You mean approval this pr? That's Okay.

Reply:
When talking system complexity, the technique "encapsulation" could be utilized. While wandering through the codebase of `mercury`, I could hardly find a struct, of which the fields are not exposed with`pub`. Not a good programming paradigm in my opinion.

Reply:
Rust's design philosophy differs from traditional object-oriented languages (e.g. Java). Exposing fields directly is an efficient and idiomatic approach that provides zero-cost abstraction.

Reply:
The official Rust API [guidelines ](https://rust-chinese-translation.github.io/api-guidelines/future-proofing.html)recommend that for composite, PDS (Passive Data Structures), such as C-style structs, exposing fields directly is a suitable approach. Git objects are an example of such structures.

Reply:
> Exposing fields directly is an efficient and idiomatic approach that provides zero-cost abstraction.

Exposing fields is appearently not abstraction, it's specialization...

when we are talking about zero-cost abstraction, we are saying those traits, generics and derives

Reply:
Yes, I was wrong about that statement. 

Directly exposing fields can reduce unnecessary abstractions. We should analyze specific problems case by case, instead of blindly following rules.

Reply:
Take a simple example.
``` Rust
pub struct Blob {
    pub id: SHA1,
    pub data: Vec<u8>,
}
```
The `id` is closely linked to the `data`, and any change in one will lead to a change in the other, which does not conform to the definition of PDS. I see no use case when you modify a existing `Blob` instead of creating a new one.

Reply:
`Blob` objects do not provide any methods to modify fields, so they conform to the definition of PDS. 

According to your logic, all pub fields can be freely modified by users, then pub fields should be banned worldwide.

Reply:
> According to your logic, all pub fields can be freely modified by users, then pub fields should be banned worldwide.

This is false dilemma.

> Blob objects do not provide any methods to modify fields, so they conform to the definition of PDS.
It indeed do not provide, but that's only because you didn't write for it...
I would implement it like this:
``` rust 
#[derive(Eq, Debug, Clone)]
pub struct Blob {
    id: SHA1,
    data: Vec<u8>,
}

impl PartialEq for Blob { ... }
impl ObjectTrait for Blob { ... }
impl Blob {
	pub fn id(&self) -> SHA1 {}
	pub fn data(&self) -> &[u8] {}
	pub fn from_foo(foo: Foo) -> Self {} // carefully checked 
}
```

By doing so, one may assert each `Blob` passed to them is **valid** because it could only be constructed from a construction function that are **guaranteed to produce a valid `Blob`**, how could this be achieved if the fields are public?

Reply:
Then, apart from the `Blob`, do you think there are issues with other Git objects?

Reply:
same for other objects contains a SHA1 hash for itself

well, visibility issue is quite common across the `mercury` codebase. It may be just about naming, but it's the first time that I see a rust library exposes its `internal` module to the outside.

Reply:
Discussing in a closed issue makes it difficult for others to follow. How about opening a new issue or starting a discussion instead? @MrBeanCpp @el-ev